client/peerset/src/protocol_controller.rs

```diff
@@ -0,0 +1,1607 @@
+			None => {
+				trace!(target: "peerset", "Adding reserved node {}.", peer_id);
+				PeerState::NotConnected
+			},
+		};
+
+		self.reserved_nodes.insert(peer_id, state.clone());
+
+		// Discount occupied slots or connect to the node.
+		match state {
+			PeerState::Connected(Direction::Inbound) => self.num_in -= 1,
+			PeerState::Connected(Direction::Outbound) => self.num_out -= 1,
+			PeerState::NotConnected => self.alloc_slots(),
+		}
+	}
+
+	/// Remove the peer from the set of reserved peers. The peer is moved to the set of regular
+	/// nodes.
+	fn on_remove_reserved_peer(&mut self, peer_id: PeerId) {
+		let mut state = match self.reserved_nodes.remove(&peer_id) {
+			Some(state) => state,
+			None => {
+				warn!(
+					target: "peerset",
+					"Trying to remove unknown reserved node: {}.",
+					peer_id,
+				);
+				return
+			},
+		};
+
+		if let PeerState::Connected(direction) = state {
+			if self.reserved_only {
+				// Disconnect the node.
+				info!(
+					target: "peerset",
+					"Disconnecting previously reserved node {} ({:?}) on {:?}.",
+					peer_id, direction, self.set_id,
+				);
+				self.drop_connection(peer_id);
+			} else {
+				// Count connections as of regular node.
+				trace!(
+					target: "peerset",
+					"Making a connected reserved node {} ({:?}) a regular one.",
+					peer_id, direction,
+				);
+				match direction {
+					Direction::Inbound => self.num_in += 1,
+					Direction::Outbound => self.num_out += 1,
+				}
+				// Put the node into the list of regular nodes.
+				let prev = self.nodes.insert(peer_id, direction);
+				assert!(prev.is_none(), "Corrupted state: reserved node was also non-reserved.");
+			}
+		} else {
+			trace!(target: "peerset", "Removed disconnected reserved node {}.", peer_id);
+		}
+	}
+
+	/// Replace the set of reserved peers.
+	fn on_set_reserved_peers(&mut self, peer_ids: HashSet<PeerId>) {
+		// Determine the difference between the current group and the new list.
+		let current = self.reserved_nodes.keys().cloned().collect();
+		let to_insert = peer_ids.difference(&current).cloned().collect::<Vec<_>>();
+		let to_remove = current.difference(&peer_ids).cloned().collect::<Vec<_>>();
+
+		for node in to_insert {
+			self.on_add_reserved_peer(node);
+		}
+
+		for node in to_remove {
+			self.on_remove_reserved_peer(node);
+		}
+	}
+
+	/// Change "reserved only" flag. In "reserved only" mode we connect and accept connections to
+	/// reserved nodes only.
+	fn on_set_reserved_only(&mut self, reserved_only: bool) {
+		trace!(target: "peerset", "Set reserved only: {}", reserved_only);
+
+		self.reserved_only = reserved_only;
+
+		if !reserved_only {
+			return self.alloc_slots()
+		}
+
+		// Disconnect all non-reserved peers.
+		self.nodes.keys().for_each(|peer_id| self.drop_connection(*peer_id));
+		self.nodes.clear();
+		self.num_in = 0;
+		self.num_out = 0;
+	}
+
+	/// Get the list of reserved peers.
+	fn on_reserved_peers(&self, pending_response: oneshot::Sender<Vec<PeerId>>) {
```

Comment:
This function name could be better. It's not really clear that it's a getter for reserved peers.

Reply:
Renamed into `on_get_reserved_peers`.